worker/provisioner/provisioner_task.go

```diff
@@ -1039,28 +1044,33 @@ func (task *provisionerTask) startMachine(
 		// find a new one to use when retrying.
 		if errors.Cause(err) == environs.ErrAvailabilityZoneFailed {
 			if !distributeAcrossZones {
-				// User specified availability zone failed, nothing to retry.
 				task.removeMachineFromAZMap(machine)
-				return task.setErrorStatus("cannot start instance for machine %q in placement zone: %v", machine, err)
-			}
-			azRemaining, err2 := task.markMachineFailedInAZ(machine, startInstanceParams.AvailabilityZone)
-			if err2 != nil {
-				if err = task.setErrorStatus("cannot start instance: %v", machine, err2); err != nil {
-					logger.Errorf("setting error status: %s", err)
+				if len(derivedZones) == 1 {
+					return task.setErrorStatus("cannot start instance for machine %q in placement zone: %v", machine, err)
 				}
-				return err2
-			}
-			if azRemaining {
-				// There are more availability zones to try, don't burn an attempt.
-				retryMsg = fmt.Sprintf("failed to start instance (%s) within attempt %d, retrying in %v with new availability zone",
-					err.Error(),
-					task.retryStartInstanceStrategy.retryCount-attemptsLeft,
-					task.retryStartInstanceStrategy.retryDelay)
-				attemptsLeft++
+				retryMsg = fmt.Sprintf("failed to to start instance (%s) in placement zone: %v, retrying in next zone", machine, err)
+				derivedZones = derivedZones[1:]
```

Comment:
This all seems a bit messy. The two use cases are very similar. Whether distributeAcrossZones is true or false, we have a list of zones to try. The list will be different, but the logic to try each one, track failures, etc should be the same. Can we refactor this logic to make it all simpler?

Reply:
refactored.